### DISCUSSION TITLE
STM32F4 with ILI9486 in 16 bit parallel help

### DISCUSSION THREAD

**Discussion by pwd666** [Bodmer/TFT_eSPI#2395] [Compatible displays and setup files]
Hello, I'm using an STM32F401 board like this: 
https://www.mischianti.org/wp-content/uploads/2022/02/STM32-STM32F4-STM32F401-STM32F401CCU6-pinout-low-resolution.jpg

with a ILI9486 display ( model MRB3503 ) like this:
http://www.lcdwiki.com/3.5inch_16BIT_Module_ILI9486_SKU:MRB3503

I am try to get the display to work in 16 bit parallel mode (that is the only supported mode by the display) but I'm just getting a white screen.

My user setup looks like:
```
#define USER_SETUP_INFO "User_Setup"
#define TFT_PARALLEL_16_BIT 
#define ILI9486_DRIVER

 #define TFT_BL   32            // LED back-light control pin
 #define TFT_BACKLIGHT_ON HIGH  // Level to turn ON back-light (HIGH or LOW)

#define TFT_CS   4  // Chip select control pin (library pulls permanently low
#define TFT_DC   31  // Data Command control pin - must use a pin in the range 0-31
#define TFT_RST  38  // Reset pin, toggles on startup

#define TFT_WR    30  // Write strobe control pin - must use a pin in the range 0-31
#define TFT_RD    29  // Read strobe control pin //possibly to 3.3V ?

#define TFT_D0   10  // Must use pins in the range 0-31 for the data bus
#define TFT_D1   11  // so a single register write sets/clears all bits.
#define TFT_D2   12  // Pins can be randomly assigned, this does not affect
#define TFT_D3   13  // TFT screen update performance.
#define TFT_D4   14
#define TFT_D5   15
#define TFT_D6   16
#define TFT_D7   17
#define TFT_D8   18  // Must use pins in the range 0-31 for the data bus
#define TFT_D9   19  // so a single register write sets/clears all bits.
#define TFT_D10   20  // Pins can be randomly assigned, this does not affect
#define TFT_D11   21  // TFT screen update performance.
#define TFT_D12   25
#define TFT_D13   26
#define TFT_D14   27
#define TFT_D15   28

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts
#define SMOOTH_FONT


#define SPI_FREQUENCY  27000000
#define SPI_TOUCH_FREQUENCY  2500000
```

On the ReadWrite_Test I get:

Pixel value written = 1
 Pixel value read    = FFFF
 ERROR                 ^^^^
 Pixel value written = 2
 Pixel value read    = FFFF
 ERROR                 ^^^^
 Pixel value written = 4
 Pixel value read    = FFFF
 ERROR                 ^^^^
 Pixel value written = 8
 Pixel value read    = FFFF
 ERROR                 ^^^^
 Pixel value written = 10
 Pixel value read    = FFFF
 ERROR                 ^^^^

On the Read_User_Setup sketch I get:

[code]
TFT_eSPI ver = 2.5.0
Processor    = Generic
Transactions = Yes
Interface    = Parallel
Display driver = 9486
Display width  = 320
Display height = 480

TFT_CS   = GPIO -1
TFT_DC   = GPIO -1
TFT_RST  = GPIO -1
TFT_WR   = GPIO -1
TFT_RD   = GPIO -1

TFT_D0   = GPIO -1
TFT_D1   = GPIO -1
TFT_D2   = GPIO -1
TFT_D3   = GPIO -1
TFT_D4   = GPIO -1
TFT_D5   = GPIO -1
TFT_D6   = GPIO -1
TFT_D7   = GPIO -1

TFT_BL           = GPIO -1
TFT_BACKLIGHT_ON = HIGH

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

[/code]

**Comment by Bodmer** [comment]
The library does not support a 16 bit parallel interface with STM32 bit processor. The library only supports 16 bit parallel with RP2040 processors. You will also get a better performance with an RP2040.

I have updated the ReadMe with a table to make this clearer.

**Reply by Bodmer** [reply]
You may be able to get the display working with an 8 bit interface but the colours will be wrong, add these lines

#define TFT_PARALLEL_8_BIT
#define STM32

 to your setup file.

**Reply by brightproject** [reply]
> You may be able to get the display working with an 8 bit interface but the colours will be wrong, add these lines
> 
> #define TFT_PARALLEL_8_BIT #define STM32
> 
> to your setup file.

Tell me, is 8-bit for stm32f401ccu6(blackpill) supported by the library?
Connected blackpill as bluepill:
```
//LCD pins  |D7 |D6 |D5 |D4 |D3 |D2 |D1 |D0 | |RD |WR |RS |CS |RST| |SD_SS|SD_DI|SD_DO|SD_SCK|
//STM32 pin |PA7|PA6|PA5|PA4|PA3|PA2|PA1|PA0| |PB0|PB6|PB7|PB8|PB9| |PB12 |PB14  |PB15  |PB13   | **SPI2**
Is right?

**Reply by Bodmer** [reply]
Yes, that looks OK.

**Comment by pwd666** [comment]
Thanks for the reply Bodmer, I have ordered a RP2040 and will go with that. I may still end up trying trying this STM32 in 8 bit mode,

**Comment by pwd666** [comment]
Just wanted to give an update. I am now using this library with a Raspberry Pie Pico and it is working great with this library. Also, I was making a very fundamental mistake when trying to get it to work: I was doing "#define TFT_D0   10" where 10 was the pin number on the board pin diagram and not the GPIO number! Hopefully someone just starting out can learn from my mistake.

Thanks for the help Bodmer, you rock.
